### ISSUE TITLE
Do not bubble Errors from Put Effects

### ISSUE THREAD

**Issue by michaelgilley** [yelouafi/redux-saga#550]
I came across an issue where a reducer was expecting a property on action data that wasn't there and so it threw an error but whenever this occurs already inside the catch statement of a saga it causes that saga, and its parents, and subsequently their children to all abort(!). This led me to question why Put effects are bubbling up errors to begin with since (it seems to me) the effect should act like an event emitter on the Redux store bus. If something throws downstream shouldn't it be captured and logged but not bubble back into the saga?

**Example:**
```js
function * watcherSaga () {
  yield * takeLatest(FOO, workerSaga)
}

function * workerSaga (action) {
  try {
    const res = yield call(xhr, {foo: action.data.foo})
    yield put(actions.fooSuccessful(res.body)
  } catch (err) {
    // uncaught thrown error here will cause the entire app to crash!
    yield put(actions.fooFailure(err)
  }
}

export default reducer ({type, data}) {
  switch (type) {
    case FOO_FAILURE:
      return state.merge({
        // this will throw and flow back into the workerSaga
        foo: data.oops.not.here
      })
    default:
      state
  }
}
```

I propose that the [runPutEffect](https://github.com/yelouafi/redux-saga/blob/master/src/internal/proc.js#L443) function handle errors much like the [Cancel Effect](https://github.com/yelouafi/redux-saga/blob/master/src/internal/proc.js#L443) by logging out any errors but returning the cb without the error. While the reducer *should* handle these error cases on its own it would still be a good defensive check to keep the entire app from cancelling.

Thoughts?

**Comment by michaelgilley** [comment]
@yelouafi Any thoughts on this?

**Comment by yelouafi** [comment]
Yeah it makes sense if the effect is non blocking.

OTOH, what could be the alternative path for an app (other than crashing)?

**Comment by michaelgilley** [comment]
If other parts of the app do throw that will indeed create problems. However, when using a more complex collection of combined reducers only the sub-reducer is failing so there is some grace there. So of course the issue that is causing an error needs addressing but the referenced PR makes the difference between a portion of the app failing and the entire saga tree falling down.

**Comment by slorber** [comment]
Agree that error in action listeners of store (middleware/reducer/other...) should not crash sagas

Inspired from backend distributed systems, when a message is published to the log/bus, it can't be removed. Other systems consume the message from the log/bus, and if they fail, it should have no impact on the system that emitted the message in the first place (in this case the saga that did the put effect)

**Comment by neurosnap** [comment]
I'm not sure if this is related or not, but when a saga fails in my current application, my sagas do not recover at all, rendering my application unusable.

**Comment by michaelgilley** [comment]
@neurosnap it is loosely connected but having some sort of recovery/retry for forks or spawns might not be a bad idea.

**Comment by slorber** [comment]
@neurosnap you have to handle that yourself, and be sure to understand that a forked saga that fails, also kill the parent. Use spawn instead to better isolate the sagas (but you'll have to handle cancellation manually)

**Comment by slorber** [comment]
@neurosnap @michaelgilley I've just created an issue for that: https://github.com/yelouafi/redux-saga/issues/570

**Comment by yelouafi** [comment]
@michaelgilley should we consider this fixed by #567
